### ISSUE TITLE
New Diagrams and Questions

### ISSUE THREAD

**Issue by leonardoaraujosantos** [weiliu89/caffe#392]
## New diagrams and layers explanation
Idea: Explain SSD flow using diagrams and describing the it's new layers (Questions on what I did not understand yet bellow)

### Training Diagram
![ssd_blockdiagram_train](https://cloud.githubusercontent.com/assets/898383/21979894/c85401c2-dbd8-11e6-80e0-1bb509fe6288.png)

### Prediction Diagram
![ssd_blockdiagram](https://cloud.githubusercontent.com/assets/898383/21979899/ce8cd2b2-dbd8-11e6-89d3-5da562e4a86b.png)

### Extra Convs
Group of 1x1 and 3x3 convs used to sample activations from CONV_4_3 and last layers (FC), it's used to improve the detection of bigger objects. 
![extraconvs](https://cloud.githubusercontent.com/assets/898383/21978172/4c7c9818-dbd1-11e6-95bc-4708bfe74059.png)

### Detection layer
![ssd_detection](https://cloud.githubusercontent.com/assets/898383/21978228/993a5cb2-dbd1-11e6-9588-6c51d1b8187c.png)

#### [Priorboxes](https://github.com/weiliu89/caffe/blob/ssd/src/caffe/layers/prior_box_layer.cpp)
Generate default boxes using the Image and feature map dimensions

#### [Annotated Data](https://github.com/weiliu89/caffe/blob/ssd/src/caffe/layers/annotated_data_layer.cpp)
1. Do image augmentation.
2. Generate random images (This is the Negative mining part??)

#### [DetectionOutput](https://github.com/weiliu89/caffe/blob/ssd/src/caffe/layers/detection_output_layer.cpp)
Do the Non-maxima supression during prediction to filter the best region per object.

#### [SmoothL1](https://github.com/weiliu89/caffe/blob/ssd/src/caffe/layers/smooth_L1_loss_layer.cpp)
Distance metric used on MultiboxLoss layer, more specifically on the Location part.

#### [Permute](https://github.com/weiliu89/caffe/blob/ssd/src/caffe/layers/permute_layer.cpp)
Used to change the dimension position on the tensors. (Don't know why??)

#### [Normalize](https://github.com/weiliu89/caffe/blob/ssd/src/caffe/layers/normalize_layer.cpp)
Used to make activations of conv4_3 smaller, (There are no tests to check if this is needed on other layers, or if they can be substituted by batchNorm or other variantes of batchNorm(NIPS2016)

#### [Deconvolution/AtrousConvolution/Dilated Conv](https://github.com/weiliu89/caffe/blob/ssd/src/caffe/layers/deconv_layer.cpp#L32)
Acutally it was alredy on caffe, during the experiemnts was found better performance (frame rate)

## Open Questions
1. Is the AnnotatedData layer responsible to augment the negative examples?
2. The normalize layer could be replaced by BatchNorm, or other variants like [WeightNorm (NIPS 2016)](https://papers.nips.cc/paper/6114-weight-normalization-a-simple-reparameterization-to-accelerate-training-of-deep-neural-networks.pdf). The idea is that during backprop we will find optimal Normalization coeficients for every detection branch
3. To handle bigger objects you used priorboxes with different feature map sizes. Does the priorbox default box sizes must be the same on all layers in order to grab a bigger portion of the feature map size or there is some kind of resizing?
4. Why we need to permute the tensors?

**Comment by weiliu89** [comment]
1. The AnnotatedData layer is only responsible to generate random patches according to the configuration. If batch_sampler allows, it is possible to generate an image without any ground truth objects. The negative mining is done in the MultiBoxLoss layer after matching ground truth boxes and default boxes.

2. Since VGG is not trained with batch norm, we found using L2 normalization (from ParseNet) is a nice and easy work around to stabilize the training with VGG. An advantage over batch normalization is that L2 normalization can be done separately per feature map location and does not depends on the number of image in a batch.

3. The size of default boxes at each feature map layer is configurable. Notice, the default boxes are in the normalized scale [0, 1].

4. To better combine predictions from different layer
